@codex Help resolve these feature test failures:

There were 3 errors:

1) Tests\Feature\OfficeAddressOperationLoggingTest::testAddressChangeCreatesStructuredOperationRecord
ErrorException: Attempt to read property "name" on null

/home/runner/work/cbdb-online-main-server/cbdb-online-main-server/vendor/laravel/framework/src/Illuminate/Foundation/Bootstrap/HandleExceptions.php:255
/home/runner/work/cbdb-online-main-server/cbdb-online-main-server/app/Repositories/BiogMainRepository.php:624
/home/runner/work/cbdb-online-main-server/cbdb-online-main-server/vendor/laravel/framework/src/Illuminate/Database/Concerns/ManagesTransactions.php:30
/home/runner/work/cbdb-online-main-server/cbdb-online-main-server/vendor/laravel/framework/src/Illuminate/Database/DatabaseManager.php:509
/home/runner/work/cbdb-online-main-server/cbdb-online-main-server/vendor/laravel/framework/src/Illuminate/Support/Facades/Facade.php:355
/home/runner/work/cbdb-online-main-server/cbdb-online-main-server/app/Repositories/BiogMainRepository.php:595
/home/runner/work/cbdb-online-main-server/cbdb-online-main-server/tests/Feature/OfficeAddressOperationLoggingTest.php:97

2) Tests\Feature\OfficeAddressOperationLoggingTest::testAddressChangeOnlyTouchesTargetOffice
ErrorException: Attempt to read property "name" on null

/home/runner/work/cbdb-online-main-server/cbdb-online-main-server/vendor/laravel/framework/src/Illuminate/Foundation/Bootstrap/HandleExceptions.php:255
/home/runner/work/cbdb-online-main-server/cbdb-online-main-server/app/Repositories/BiogMainRepository.php:624
/home/runner/work/cbdb-online-main-server/cbdb-online-main-server/vendor/laravel/framework/src/Illuminate/Database/Concerns/ManagesTransactions.php:30
/home/runner/work/cbdb-online-main-server/cbdb-online-main-server/vendor/laravel/framework/src/Illuminate/Database/DatabaseManager.php:509
/home/runner/work/cbdb-online-main-server/cbdb-online-main-server/vendor/laravel/framework/src/Illuminate/Support/Facades/Facade.php:355
/home/runner/work/cbdb-online-main-server/cbdb-online-main-server/app/Repositories/BiogMainRepository.php:595
/home/runner/work/cbdb-online-main-server/cbdb-online-main-server/tests/Feature/OfficeAddressOperationLoggingTest.php:166

3) Tests\Feature\OfficeAddressOperationLoggingTest::testAddressUpdateKeepsOtherOfficeRows
ErrorException: Attempt to read property "name" on null

/home/runner/work/cbdb-online-main-server/cbdb-online-main-server/vendor/laravel/framework/src/Illuminate/Foundation/Bootstrap/HandleExceptions.php:255
/home/runner/work/cbdb-online-main-server/cbdb-online-main-server/app/Repositories/BiogMainRepository.php:624
/home/runner/work/cbdb-online-main-server/cbdb-online-main-server/vendor/laravel/framework/src/Illuminate/Database/Concerns/ManagesTransactions.php:30
/home/runner/work/cbdb-online-main-server/cbdb-online-main-server/vendor/laravel/framework/src/Illuminate/Database/DatabaseManager.php:509
/home/runner/work/cbdb-online-main-server/cbdb-online-main-server/vendor/laravel/framework/src/Illuminate/Support/Facades/Facade.php:355
/home/runner/work/cbdb-online-main-server/cbdb-online-main-server/app/Repositories/BiogMainRepository.php:595
/home/runner/work/cbdb-online-main-server/cbdb-online-main-server/tests/Feature/OfficeAddressOperationLoggingTest.php:235

ERRORS!
Tests: 249, Assertions: 855, Errors: 3.
Error: Process completed with exit code 2.

--- a/tests/Feature/BasicInformationSourcesControllerTest.php
+++ b/tests/Feature/BasicInformationSourcesControllerTest.php
@@ -1,0 +1,181 @@
+<?php
+
+namespace Tests\Feature;
+
+use App\User;
+use App\Repositories\BiogMainRepository;
+use Illuminate\Database\Schema\Blueprint;
+use Illuminate\Http\Request;
+use Illuminate\Support\Facades\Auth;
+use Illuminate\Support\Facades\Config;
+use Illuminate\Support\Facades\DB;
+use Illuminate\Support\Facades\Schema;
+use Tests\TestCase;
+
+class BasicInformationSourcesControllerTest extends TestCase
+{
+    protected function setUp(): void
+    {
+        parent::setUp();
+
+        if (!extension_loaded('pdo_sqlite')) {
+            $this->markTestSkipped('pdo_sqlite extension is required for this test.');
+        }
+
+        Config::set('database.default', 'sqlite');
+        Config::set('database.connections.sqlite', [
+            'driver' => 'sqlite',
+            'database' => ':memory:',
+            'prefix' => '',
+        ]);
+
+        DB::purge('sqlite');
+        DB::reconnect('sqlite');
+
+        Schema::create('BIOG_SOURCE_DATA', function (Blueprint $table) {
+            $table->integer('c_personid');
+            $table->integer('c_textid');
+            $table->string('c_pages');
+            $table->text('c_notes')->nullable();
+            $table->tinyInteger('c_main_source')->default(0);
+            $table->tinyInteger('c_self_bio')->default(0);
+            $table->string('c_created_by')->nullable();
+            $table->timestamp('c_created_date')->nullable();
+            $table->string('c_modified_by')->nullable();
+            $table->timestamp('c_modified_date')->nullable();
+        });
+
+        Schema::create('operations', function (Blueprint $table) {
+            $table->increments('id');
+            $table->unsignedInteger('user_id')->nullable();
+            $table->integer('c_personid')->nullable();
+            $table->tinyInteger('op_type');
+            $table->string('resource');
+            $table->string('resource_id')->nullable();
+            $table->text('resource_data')->nullable();
+            $table->text('resource_original')->nullable();
+            $table->timestamps();
+            $table->tinyInteger('crowdsourcing_status')->default(0);
+            $table->tinyInteger('rate')->default(0);
+        });
+
+        Auth::guard()->setUser($this->makeUser(9, 'Creator User'));
+    }
+
+    protected function makeUser(int $id, string $name): User
+    {
+        $user = new User();
+        $user->id = $id;
+        $user->name = $name;
+        $user->avatar = 'avatar5.png';
+        $user->is_admin = User::ROLE_SUPER_ADMIN;
+        $user->is_active = User::STATUS_ACTIVE;
+
+        return $user;
+    }
+
+    public function testSourceStoreWritesAuditFieldsAndOperations(): void
+    {
+        $repository = new BiogMainRepository();
+        $request = new Request([
+            'c_textid' => 500,
+            'c_pages' => 'p10',
+            'c_notes' => 'first note',
+            'c_main_source' => 1,
+            'c_self_bio' => 0,
+        ]);
+
+        $result = $repository->sourceStoreById($request, 321);
+
+        $this->assertSame('Creator User', $result['c_created_by']);
+        $this->assertNotNull($result['c_created_date']);
+
+        $row = DB::table('BIOG_SOURCE_DATA')->first();
+        $this->assertSame('Creator User', $row->c_created_by);
+        $this->assertNotNull($row->c_created_date);
+        $this->assertNull($row->c_modified_by);
+        $this->assertNull($row->c_modified_date);
+
+        $operation = DB::table('operations')->first();
+        $this->assertSame('BIOG_SOURCE_DATA', $operation->resource);
+        $this->assertSame('321-500-p10', $operation->resource_id);
+        $this->assertEquals(1, $operation->op_type);
+        $this->assertNull($operation->resource_original);
+
+        $payload = json_decode($operation->resource_data, true);
+        $this->assertSame('Creator User', $payload['c_created_by']);
+        $this->assertSame(321, $payload['c_personid']);
+    }
+
+    public function testSourceUpdatePreservesCreationAndSetsModification(): void
+    {
+        $repository = new BiogMainRepository();
+        $initialRequest = new Request([
+            'c_textid' => 501,
+            'c_pages' => 'p20',
+            'c_notes' => 'initial note',
+            'c_main_source' => 0,
+            'c_self_bio' => 1,
+        ]);
+        $repository->sourceStoreById($initialRequest, 654);
+
+        $original = DB::table('BIOG_SOURCE_DATA')->first();
+
+        Auth::guard()->setUser($this->makeUser(10, 'Editor User'));
+
+        $updateRequest = new Request([
+            'c_textid' => 501,
+            'c_pages' => 'p20',
+            'c_notes' => 'updated note',
+            'c_main_source' => 1,
+            'c_self_bio' => 0,
+        ]);
+
+        $repository->sourceUpdateById($updateRequest, 654, '654-501-p20');
+
+        $row = DB::table('BIOG_SOURCE_DATA')->first();
+        $this->assertSame('Creator User', $row->c_created_by);
+        $this->assertSame($original->c_created_date, $row->c_created_date);
+        $this->assertSame('Editor User', $row->c_modified_by);
+        $this->assertNotNull($row->c_modified_date);
+
+        $operation = DB::table('operations')->orderByDesc('id')->first();
+        $this->assertSame('BIOG_SOURCE_DATA', $operation->resource);
+        $this->assertSame('654-501-p20', $operation->resource_id);
+        $this->assertEquals(3, $operation->op_type);
+        $this->assertNull($operation->resource_original);
+
+        $payload = json_decode($operation->resource_data, true);
+        $this->assertSame('Editor User', $payload['c_modified_by']);
+        $this->assertSame('updated note', $payload['c_notes']);
+    }
+
+    public function testEditViewDisplaysCreationAndModificationInfo(): void
+    {
+        $row = (object) [
+            'c_personid' => 777,
+            'c_textid' => 888,
+            'c_pages' => 'A-1',
+            'c_notes' => 'view-note',
+            'c_main_source' => 1,
+            'c_self_bio' => 0,
+            'c_created_by' => 'Creator User',
+            'c_created_date' => '2024-02-01 12:00:00',
+            'c_modified_by' => 'Editor User',
+            'c_modified_date' => '2024-02-03 08:30:00',
+        ];
+
+        $html = view('biogmains.sources.edit', [
+            'id' => 777,
+            'row' => $row,
+            'res' => ['text_str' => 'Dummy Text'],
+            'page_title' => 'Basicinformation',
+            'page_description' => '基本信息表 出處',
+            'page_url' => '/basicinformation/777/sources',
+            'archer' => "<li><a href='#'>Sources</a></li>",
+        ])->render();
+
+        $this->assertStringContainsString('Creator User/2024-02-01 12:00:00', $html);
+        $this->assertStringContainsString('Editor User/2024-02-03 08:30:00', $html);
+    }
+}

--- a/tests/Feature/OfficeAddressOperationLoggingTest.php
+++ b/tests/Feature/OfficeAddressOperationLoggingTest.php
@@ -3,12 +3,14 @@
 namespace Tests\Feature;
 
 use App\Repositories\BiogMainRepository;
+use Illuminate\Auth\GenericUser;
 use Illuminate\Http\Request;
 
 use Illuminate\Support\Facades\Config;
 use Illuminate\Support\Facades\DB;
 use Illuminate\Support\Facades\Schema;
 use Illuminate\Database\Schema\Blueprint;
+use Illuminate\Support\Facades\Auth;
 use Tests\TestCase;
 
 class OfficeAddressOperationLoggingTest extends TestCase
@@ -47,6 +49,19 @@ class OfficeAddressOperationLoggingTest extends TestCase
             $table->integer('c_posting_id');
             $table->integer('c_office_id');
             $table->integer('c_addr_id');
+            $table->string('c_created_by')->nullable();
+            $table->timestamp('c_created_date')->nullable();
+            $table->string('c_modified_by')->nullable();
+            $table->timestamp('c_modified_date')->nullable();
+        });
+
+        Schema::create('POSTING_DATA', function (Blueprint $table) {
+            $table->integer('c_personid');
+            $table->integer('c_posting_id')->primary();
+            $table->string('c_created_by')->nullable();
+            $table->timestamp('c_created_date')->nullable();
+            $table->string('c_modified_by')->nullable();
+            $table->timestamp('c_modified_date')->nullable();
         });
 
         Schema::create('operations', function (Blueprint $table) {
@@ -62,6 +77,8 @@ class OfficeAddressOperationLoggingTest extends TestCase
             $table->tinyInteger('crowdsourcing_status')->default(0);
             $table->tinyInteger('rate')->default(0);
         });
+
+        Auth::guard()->setUser(new GenericUser(['id' => 1, 'name' => 'Testing Admin']));
     }
 
     public function testAddressChangeCreatesStructuredOperationRecord(): void

--- a/tests/Feature/OfficeAddressOperationLoggingTest.php
+++ b/tests/Feature/OfficeAddressOperationLoggingTest.php
@@ -117,6 +117,7 @@ class OfficeAddressOperationLoggingTest extends TestCase
         $this->assertNotNull($operation);
         $this->assertEquals('POSTED_TO_ADDR_DATA', $operation->resource);
         $this->assertEquals('71313-312754', $operation->resource_id);
+        $this->assertNotNull($operation->resource_original);
 
         $after = json_decode($operation->resource_data, true);
         $before = json_decode($operation->resource_original, true);
@@ -142,6 +143,137 @@ class OfficeAddressOperationLoggingTest extends TestCase
                 ],
             ],
         ], $before);
+    }
+
+    public function testAddressChangeUpdatesAuditFields(): void
+    {
+        DB::table('POSTING_DATA')->insert([
+            'c_personid' => 900001,
+            'c_posting_id' => 880001,
+            'c_created_by' => 'Seeder',
+            'c_created_date' => '2023-01-01 00:00:00',
+        ]);
+
+        DB::table('POSTED_TO_OFFICE_DATA')->insert([
+            'c_personid' => 900001,
+            'c_office_id' => 77771,
+            'c_posting_id' => 880001,
+            'c_fy_intercalary' => 0,
+            'c_ly_intercalary' => 0,
+            'c_source' => 0,
+        ]);
+
+        DB::table('POSTED_TO_ADDR_DATA')->insert([
+            'c_personid' => 900001,
+            'c_posting_id' => 880001,
+            'c_office_id' => 77771,
+            'c_addr_id' => 45,
+            'c_created_by' => 'Seeder',
+            'c_created_date' => '2023-01-02 00:00:00',
+        ]);
+
+        Auth::guard()->setUser(new GenericUser(['id' => 2, 'name' => 'Updater A']));
+
+        $request = new Request([
+            '_id' => 900001,
+            '_postingid' => 880001,
+            '_officeid' => 77771,
+            'c_office_id' => 77771,
+            'c_fy_intercalary' => 0,
+            'c_ly_intercalary' => 0,
+            'c_source' => 0,
+            'c_addr' => [46],
+        ]);
+
+        $repository = new BiogMainRepository();
+        $repository->officeUpdateById($request, 880001, 900001);
+
+        $posting = DB::table('POSTING_DATA')->where('c_posting_id', 880001)->first();
+        $this->assertSame('Updater A', $posting->c_modified_by);
+        $this->assertNotNull($posting->c_modified_date);
+
+        $address = DB::table('POSTED_TO_ADDR_DATA')->where([
+            'c_personid' => 900001,
+            'c_posting_id' => 880001,
+            'c_office_id' => 77771,
+            'c_addr_id' => 46,
+        ])->first();
+
+        $this->assertSame('Updater A', $address->c_created_by);
+        $this->assertSame('Updater A', $address->c_modified_by);
+        $this->assertNotNull($address->c_created_date);
+        $this->assertNotNull($address->c_modified_date);
+
+        $operation = DB::table('operations')->orderByDesc('id')->first();
+        $this->assertSame('POSTED_TO_ADDR_DATA', $operation->resource);
+        $this->assertNotNull($operation->resource_original);
+        $payload = json_decode($operation->resource_data, true);
+        $original = json_decode($operation->resource_original, true);
+
+        $this->assertSame(46, $payload['rows'][0]['c_addr_id']);
+        $this->assertSame(45, $original['rows'][0]['c_addr_id']);
+    }
+
+    public function testOfficeUpdateDoesNotResetAddressTimestampsWhenUnchanged(): void
+    {
+        DB::table('POSTING_DATA')->insert([
+            'c_personid' => 910002,
+            'c_posting_id' => 880002,
+            'c_created_by' => 'Seeder',
+            'c_created_date' => '2023-05-01 00:00:00',
+            'c_modified_by' => 'Seeder',
+            'c_modified_date' => '2023-05-02 00:00:00',
+        ]);
+
+        DB::table('POSTED_TO_OFFICE_DATA')->insert([
+            'c_personid' => 910002,
+            'c_office_id' => 77772,
+            'c_posting_id' => 880002,
+            'c_fy_intercalary' => 0,
+            'c_ly_intercalary' => 0,
+            'c_source' => 0,
+        ]);
+
+        DB::table('POSTED_TO_ADDR_DATA')->insert([
+            'c_personid' => 910002,
+            'c_posting_id' => 880002,
+            'c_office_id' => 77772,
+            'c_addr_id' => 25,
+            'c_created_by' => 'Seeder',
+            'c_created_date' => '2023-05-01 00:00:00',
+            'c_modified_by' => 'Seeder',
+            'c_modified_date' => '2023-05-02 00:00:00',
+        ]);
+
+        Auth::guard()->setUser(new GenericUser(['id' => 3, 'name' => 'Updater B']));
+
+        $request = new Request([
+            '_id' => 910002,
+            '_postingid' => 880002,
+            '_officeid' => 77772,
+            'c_office_id' => 77772,
+            'c_fy_intercalary' => 1,
+            'c_ly_intercalary' => 0,
+            'c_source' => 0,
+            'c_addr' => [25],
+        ]);
+
+        $repository = new BiogMainRepository();
+        $repository->officeUpdateById($request, 880002, 910002);
+
+        $address = DB::table('POSTED_TO_ADDR_DATA')->where([
+            'c_personid' => 910002,
+            'c_posting_id' => 880002,
+            'c_office_id' => 77772,
+            'c_addr_id' => 25,
+        ])->first();
+
+        $this->assertSame('2023-05-02 00:00:00', $address->c_modified_date);
+        $this->assertSame('Seeder', $address->c_modified_by);
+
+        $operations = DB::table('operations')->pluck('resource');
+        $this->assertContains('POSTED_TO_OFFICE_DATA', $operations);
+        $this->assertNotContains('POSTED_TO_ADDR_DATA', $operations);
     }
 
     public function testAddressChangeOnlyTouchesTargetOffice(): void

